universal.c

```diff
@@ -593,10 +593,13 @@ XS(XS_utf8_upgrade)
         croak_xs_usage(cv, "sv");
     else {
         SV * const sv = ST(0);
-        STRLEN	RETVAL;
+        STRLEN	RETVAL = 0;
         dXSTARG;
 
-        RETVAL = sv_utf8_upgrade(sv);
+        if (LIKELY(sv && SvANY(sv))) {
```

Comment:
I would have said that this should be `sv && SvOK(sv) && !SvROK(sv)`, but it might be that `SvANY(sv)` has the same result. I think i would prefer we check the flags. :-)

Reply:
I think i might make it return undef if it does nothing.

Reply:
Maybe it should just be `sv && SvOK(sv)`, eg, consider overloaded refs, but including `SvROK(sv)` helped me find what are arguably bugs in our test suite. (calling it on a qr// object).

Reply:
FWIW, not knowing any better, I  copied the template from pp_defined

Reply:
Expect it to look like this:

```
diff --git a/universal.c b/universal.c
index 0b12a231b5..931487364e 100644
--- a/universal.c
+++ b/universal.c
@@ -596,11 +596,14 @@ XS(XS_utf8_upgrade)
         STRLEN RETVAL = 0;
         dXSTARG;
 
-        if (LIKELY(sv && SvANY(sv))) {
+        XSprePUSH;
+        if (LIKELY(sv && SvOK(sv))) {
             RETVAL = sv_utf8_upgrade(sv);
+            PUSHi((IV)RETVAL);
+        } else {
+            PUSHs(newSV(0));
         }
 
-        XSprePUSH; PUSHi((IV)RETVAL);
     }
     XSRETURN(1);
 }
```

Eg, it should return undef if the argument is undef. I dont think testing SvANY() is correct, it should just test SvOK().

```
    "SvOK"
        Returns a U32 value indicating whether the value is defined. This is
        only meaningful for scalars.

            U32  SvOK(SV* sv)
```

The `SvANY`  in pp_defined is a bit surprising to me, maybe its faster? Weird. Why wouldnt SvOK() be implemented that way if its faster?  I know i have seen code that basically turns off the SvOK() flag to "undef it" temporarily.

Anyway, we definitely dont want to test `SvROK()` like i thought at first. We have a bunch of regex test code that expects to be able to upgrade the pattern. It just happens to work, but it means our tests double compile the pattern.

Reply:
Also see https://github.com/Perl/perl5/pull/20453 which I found by applying the above patch but also including `!SvROK(sv)` in the guard clause.

Reply:
BTW, I reviewed pp_defined again, and I think you have misread the code it contains. SvANY() as a check is *definitely* not sufficient to determine if an SV is undef (and I question its use in pp_defined at all actually). It checks to the see if the SV has something in the any_ptr slot. Any SV that does not cannot be defined, but an SV that does may still not be defined. If you review the logic further it falls through to a bunch of logic that is meant to handle the different types, which I think is why it does the SvANY check early so that those checks are "safe", and then it falls through to further logic which finally does the SvOK().

Since your code is dealing only with SV's (and not CV's or AV's or what not, as you can not pass such from perl to XS directly, they have to come in SV (RV) wrapper) at the XS level the correct way to see if an SV is defined is to simply call SvOK(), the SvANY check is not sufficient, and unnecessary if you have to call SvOK() anyway.

I have to say I find the code in defined() fairly odd, i would have expected a fairly minimal wrapper around SvOK().  But I am 100% certain it is not the code to copy for your use case.